### DISCUSSION TITLE
Sharing a request id between Apache logs and Mojolicious logs

### DISCUSSION THREAD

**Discussion by robrwo** [mojolicious/mojo#1848] [Show and tell]
Apache has the ability to generate a unique id for each request, and include it in the logs.  This request id can be saved in your Apache logs and passed to a Mojolicious application behind a reverse proxy, so that you can connect a specific request in your application logs with a request in your Apache logs.

In Apache configuration, enable `mod_unique_id`
```apache
LoadModule unique_id_module modules/mod_unique_id.so
```
then add a custom log that includes the unique id:
```apache
LogFormat "%h %l %u %t \"%r\" %>s %b \"%{Referer}i\" \"%{User-agent}i\" %{UNIQUE_ID}e" combined_id

...

CustomLog logs/access.log combined_id
```

In your reverse proxy configuration, add a header with the unique id:
```apache
  ProxyPass http://localhost:8080 retry=0
  ...
  RequestHeader set X-UID "%{UNIQUE_ID}e"
  ProxyPassReverse http://localhost:8082
```

Then update your Mojolicious application to use the request id from Apache:
```perl
$self->hook(
    before_dispatch => sub ($c) {
        if ( my $id = $c->req->headers->header('X-UID') ) {
            $c->req->request_id($id);
        }
    }
);
```

**Comment by s1037989** [comment]
This is awesome. I use nginx. I'll need to investigate doing something similar for that, and your contributions should provide some great direction for how to possibly go about it.

But, more importantly, thanks for sharing the idea and implementation on the discussion boards. It's really great to see discussion like this here!
